src/shopify/graphql/fetchProduct.ts

```diff
@@ -1,26 +1,108 @@
 import { flattenResponse } from "./utils"
-import getProductByHandle from "@/shopify/graphql/getProductByHandle.graphql?raw"
+import getProductsByHandles from "@/shopify/graphql/getProductsByHandles.graphql?raw"
 import { getApiUrl } from "./constants"
 import { cached } from "@/utils/cached"
+import { ShopifyProduct } from "./types"
+
+type PendingRequest = {
+  handle: string
+  resolve: (product: ShopifyProduct) => void
+  reject: (error: Error) => void
+}
+
+const state = {
+  pendingRequests: new Map<string, PendingRequest[]>(),
+  scheduledFlush: null as number | null
+}
+
+async function fetchBatch(handles: string[], requestsMap: Map<string, PendingRequest[]>) {
+  // Build query string to match products by handle
+  const queryString = handles.map(h => `handle:${h}`).join(" OR ")
 
-export const [fetchProduct, clearProductCache] = cached(async (handle: string) => {
   const response = await fetch(getApiUrl().href, {
     headers: {
       "Content-Type": "application/json"
     },
     method: "POST",
     body: JSON.stringify({
-      query: getProductByHandle,
+      query: getProductsByHandles,
       variables: {
         language: window.Shopify?.locale?.toUpperCase() || "EN",
         country: window.Shopify?.country || "US",
-        handle
+        query: queryString,
+        first: handles.length
       }
     })
   })
+
   if (!response.ok) {
-    throw new Error(`Failed to fetch product data: ${response.status} ${response.statusText}`)
+    const error = new Error(`Failed to fetch product data: ${response.status} ${response.statusText}`)
+    Array.from(requestsMap.values())
+      .flatMap(requests => requests)
+      .forEach(request => request.reject(error))
+    return
   }
+
   const responseData = await response.json()
-  return flattenResponse(responseData)
-})
+
+  // Create a map of handle to product
+  const productsByHandle = new Map<string, ShopifyProduct>()
+  if (responseData.data?.products?.nodes) {
+    responseData.data.products.nodes.forEach((productNode: { handle: string }) => {
+      const product = flattenResponse({ data: { product: productNode } })
+      productsByHandle.set(productNode.handle, product)
```

Comment:
add the to the map also based on all of the variant.product.onlineStoreUrl derived handles